File: server/internal/qwk/qwk.go
// Package qwk implements the classic QWK offline-mail packet format (and its
// .REP reply-packet counterpart) using only the standard library. A .QWK packet
// is a ZIP archive a caller downloads to read a board's messages offline; a .REP
// packet is the ZIP they upload with their offline replies.
//
// The format is faithful to the de-facto QWK spec used by Qmail/QWKE-era
// readers: a text CONTROL.DAT describing the board and its conferences, plus a
// MESSAGES.DAT of fixed 128-byte records (a header block followed by N body
// blocks per message). This file is pure logic -- no I/O beyond the in-memory
// ZIP bytes -- so it is trivially testable.
package qwk

import (
	"archive/zip"
	"bytes"
	"errors"
	"fmt"
	"io"
	"strconv"
	"strings"
	"time"
)

// BBSID is the short board identifier embedded in the packet (also the .NDX /
// .MSG basename). We name the packet VENDX.QWK and expect VENDX.MSG in replies.
const BBSID = "VENDX"

// blockSize is the fixed QWK record length. MESSAGES.DAT is a sequence of these:
// record 0 is an ident block, then each message is one header block plus its
// body blocks.
const blockSize = 128

// qwkNewline is the byte QWK substitutes for a line break inside a message body
// (the classic 0xE3 / "π"); readers convert it back to a newline.
const qwkNewline = 0xE3

// Conference is one readable area in the packet, numbered for the reader.
type Conference struct {
	Number uint16
	Name   string
}

// Message is one post carried in (or recovered from) a packet.
type Message struct {
	Conference              uint16
	To, From, Subject, Body string
	Date                    time.Time
	Read                    bool
}

// Packet is everything needed to build a .QWK download for one caller.
type Packet struct {
	BoardName, Sysop, Caller string
	Conferences              []Conference
	Messages                 []Message
}

// Build renders the packet as the bytes of a .QWK ZIP archive (CONTROL.DAT +
// MESSAGES.DAT). The conferences should be the same set the messages reference.
func Build(p Packet) ([]byte, error) {
	var buf bytes.Buffer
	zw := zip.NewWriter(&buf)

	control := buildControl(p)
	if err := writeZipFile(zw, "CONTROL.DAT", control); err != nil {
		return nil, err
	}

	messages := buildMessages(p)
	if err := writeZipFile(zw, "MESSAGES.DAT", messages); err != nil {
		return nil, err
	}

	if err := zw.Close(); err != nil {
		return nil, fmt.Errorf("qwk: close zip: %w", err)
	}
	return buf.Bytes(), nil
}

// buildControl renders CONTROL.DAT: CRLF text describing the board, the caller,
// the total message count, the highest conference number, and each conference's
// number+name, then the customary blank door/news trailer lines.
func buildControl(p Packet) []byte {
	now := time.Now()
	maxConf := uint16(0)
	for _, c := range p.Conferences {
		if c.Number > maxConf {
			maxConf = c.Number
		}
	}

	var b strings.Builder
	wl := func(s string) { b.WriteString(s); b.WriteString("\r\n") }

	wl(emptyOr(p.BoardName, "Vendetta/X"))           // board name
	wl("")                                           // location (blank ok)
	wl("")                                           // phone (blank ok)
	wl(emptyOr(p.Sysop, "Sysop"))                    // sysop
	wl("00000000," + BBSID)                          // <serial>,<BBSID>
	wl(now.Format("01-02-2006,15:04:05"))            // date,time
	wl(strings.ToUpper(emptyOr(p.Caller, "CALLER"))) // caller name
	wl("")                                           // blank
	wl("")                                           // blank
	wl("0")                                          // total-messages-this-call marker
	wl(strconv.Itoa(len(p.Messages)))                // total messages in packet
	wl(strconv.Itoa(int(maxConf)))                   // highest conference number

	for _, c := range p.Conferences {
		wl(strconv.Itoa(int(c.Number)))
		wl(c.Name)
	}

	// Door / news / goodbye filenames (blank trailers are accepted).
	wl("VENDX.NWS")
	wl("HELLO")
	wl("BBSNEWS")
	wl("GOODBYE")

	return []byte(b.String())
}

// buildMessages renders MESSAGES.DAT: an ident block (record 0) followed by, for
// each message, a header block and its body blocks.
func buildMessages(p Packet) []byte {
	var out bytes.Buffer
	out.Write(identBlock("Vendetta/X QWK packet  -  Produced by Vendetta/X"))

	for i, m := range p.Messages {
		body := encodeBody(m.Body)
		bodyBlocks := (len(body) + blockSize - 1) / blockSize
		if bodyBlocks == 0 {
			bodyBlocks = 1 // a message always carries at least one body block
		}
		// Block count INCLUDING the header block.
		totalBlocks := bodyBlocks + 1

		out.Write(messageHeader(m, i+1, totalBlocks))

		// Body, padded out to a whole number of 128-byte blocks with spaces.
		padded := bodyBlocks * blockSize
		out.Write(body)
		out.Write(spaces(padded - len(body)))
	}
	return out.Bytes()
}

// identBlock returns a 128-byte ident record (record 0). Any 128-byte ident is
// accepted by readers; we use a clear Vendetta/X string, space-padded.
func identBlock(ident string) []byte {
	return fieldBytes(ident, blockSize)
}

// messageHeader returns the 128-byte header block for one message. seq is the
// 1-based logical message number within the packet; totalBlocks is the block
// count INCLUDING this header.
func messageHeader(m Message, seq, totalBlocks int) []byte {
	h := spaces(blockSize)

	status := byte(' ') // public, unread
	if m.Read {
		status = '-' // public, read
	}
	h[0] = status

	date := m.Date
	if date.IsZero() {
		date = time.Now()
	}

	putField(h, 1, 7, strconv.Itoa(seq))           // message number
	putField(h, 8, 8, date.Format("01-02-06"))     // date MM-DD-YY
	putField(h, 16, 5, date.Format("15:04"))       // time HH:MM
	putField(h, 21, 25, m.To)                      // To
	putField(h, 46, 25, m.From)                    // From
	putField(h, 71, 25, m.Subject)                 // Subject
	putField(h, 96, 12, "")                        // password (blank)
	putField(h, 108, 8, "")                        // reference msg number (blank)
	putField(h, 116, 6, strconv.Itoa(totalBlocks)) // block count incl. header

	h[122] = 0xE1 // active flag

	putUint16LE(h, 123, m.Conference) // conference number
	putUint16LE(h, 125, uint16(seq))  // logical message number in packet
	h[127] = ' '                      // net tag

	return h
}

// encodeBody converts a message body to QWK on-the-wire form: newlines become
// the 0xE3 marker so the body is a single run of bytes within its blocks.
func encodeBody(body string) []byte {
	// Normalize CRLF/CR to LF first, then map each LF to the QWK marker.
	body = strings.ReplaceAll(body, "\r\n", "\n")
	body = strings.ReplaceAll(body, "\r", "\n")
	b := []byte(body)
	for i := range b {
		if b[i] == '\n' {
			b[i] = qwkNewline
		}
	}
	return b
}

// decodeBody reverses encodeBody: 0xE3 markers become newlines and trailing
// block-padding spaces are trimmed.
func decodeBody(raw []byte) string {
	b := make([]byte, len(raw))
	copy(b, raw)
	for i := range b {
		if b[i] == qwkNewline {
			b[i] = '\n'
		}
	}
	// Block bodies are space-padded to a 128-byte boundary; trim that padding
	// (and any trailing newline introduced by it) without eating interior text.
	return strings.TrimRight(string(b), " \x00\n")
}

// ParseReply parses a .REP reply ZIP (containing VENDX.MSG in MESSAGES.DAT
// layout) and returns the caller's reply messages: conference number plus
// To/From/Subject/Body (with 0xE3 decoded back to newlines).
// maxReplyBytes caps the decompressed size of a reply packet's message member,
// guarding against zip bombs (the upload itself is also size-capped upstream).
const maxReplyBytes = 32 << 20 // 32 MiB

func ParseReply(zipBytes []byte) ([]Message, error) {
	zr, err := zip.NewReader(bytes.NewReader(zipBytes), int64(len(zipBytes)))
	if err != nil {
		return nil, fmt.Errorf("qwk: open reply zip: %w", err)
	}

	var data []byte
	wantUpper := strings.ToUpper(BBSID + ".MSG")
	for _, f := range zr.File {
		if strings.EqualFold(f.Name, BBSID+".MSG") ||
			strings.ToUpper(baseName(f.Name)) == wantUpper {
			rc, err := f.Open()
			if err != nil {
				return nil, fmt.Errorf("qwk: open %s: %w", f.Name, err)
			}
			// Bound the decompressed read so a zip bomb (a tiny member that
			// inflates to gigabytes) can't exhaust memory.
			data, err = io.ReadAll(io.LimitReader(rc, maxReplyBytes+1))
			rc.Close()
			if err != nil {
				return nil, fmt.Errorf("qwk: read %s: %w", f.Name, err)
			}
			if len(data) > maxReplyBytes {
				return nil, errors.New("qwk: reply packet member too large")
			}
			break
		}
	}
	if data == nil {
		return nil, errors.New("qwk: reply packet has no " + BBSID + ".MSG")
	}
	return parseMessages(data)
}

// parseMessages walks a MESSAGES.DAT byte stream: record 0 is the BBSID ident
// block, then each message is a header block followed by (blockCount-1) body
// blocks. Malformed/short records terminate parsing gracefully.
func parseMessages(data []byte) ([]Message, error) {
	if len(data) < blockSize {
		return nil, errors.New("qwk: message data too short")
	}
	// Skip record 0 (the ident block).
	pos := blockSize
	var out []Message

	for pos+blockSize <= len(data) {
		header := data[pos : pos+blockSize]
		pos += blockSize

		blocks := atoiField(header[116:122])
		if blocks < 1 {
			// Not a sane header; stop rather than misread the rest.
			break
		}
		bodyBlocks := blocks - 1

		bodyEnd := pos + bodyBlocks*blockSize
		if bodyEnd > len(data) {
			bodyEnd = len(data) // tolerate a truncated final block
		}
		bodyRaw := data[pos:bodyEnd]
		pos = bodyEnd

		out = append(out, Message{
			Conference: getUint16LE(header, 123),
			To:         trimField(header[21:46]),
			From:       trimField(header[46:71]),
			Subject:    trimField(header[71:96]),
			Body:       decodeBody(bodyRaw),
		})
	}
	return out, nil
}

// ---- low-level field helpers ----

// writeZipFile adds one stored file to the archive.
func writeZipFile(zw *zip.Writer, name string, data []byte) error {
	fw, err := zw.Create(name)
	if err != nil {
		return fmt.Errorf("qwk: create %s: %w", name, err)
	}
	if _, err := fw.Write(data); err != nil {
		return fmt.Errorf("qwk: write %s: %w", name, err)
	}
	return nil
}

// spaces returns an n-byte slice filled with ASCII spaces.
func spaces(n int) []byte {
	if n < 0 {
		n = 0
	}
	b := make([]byte, n)
	for i := range b {
		b[i] = ' '
	}
	return b
}

// fieldBytes returns s as exactly n bytes: truncated if too long, space-padded
// on the right if too short.
func fieldBytes(s string, n int) []byte {
	b := spaces(n)
	src := []byte(s)
	if len(src) > n {
		src = src[:n]
	}
	copy(b, src)
	return b
}

// putField writes s into dst[off:off+width], space-padded / truncated to width.
func putField(dst []byte, off, width int, s string) {
	copy(dst[off:off+width], fieldBytes(s, width))
}

// putUint16LE writes v little-endian at dst[off:off+2].
func putUint16LE(dst []byte, off int, v uint16) {
	dst[off] = byte(v)
	dst[off+1] = byte(v >> 8)
}

// getUint16LE reads a little-endian uint16 at src[off:off+2].
func getUint16LE(src []byte, off int) uint16 {
	if off+2 > len(src) {
		return 0
	}
	return uint16(src[off]) | uint16(src[off+1])<<8
}

// trimField returns a header text field with trailing/leading spaces and NULs
// removed.
func trimField(b []byte) string {
	return strings.Trim(string(b), " \x00")
}

// atoiField parses a space-padded ASCII integer field (block counts etc.).
func atoiField(b []byte) int {
	n, err := strconv.Atoi(strings.TrimSpace(string(b)))
	if err != nil {
		return 0
	}
	return n
}

// emptyOr returns s, or alt when s is blank.
func emptyOr(s, alt string) string {
	if strings.TrimSpace(s) == "" {
		return alt
	}
	return s
}

// baseName returns the final path element of a (possibly slashed) zip entry name.
func baseName(name string) string {
	if i := strings.LastIndexAny(name, "/\\"); i >= 0 {
		return name[i+1:]
	}
	return name
}

File: legacy/core/qwk.c
/*
 * qwk.c -- QWK/REP offline-mail packet FORMAT (MESSAGES.DAT + CONTROL.DAT).
 *
 * MESSAGES.DAT is a stream of 128-byte blocks:
 *   block 0   : 128-byte text header (vendor stamp + board name, space-padded)
 *   per msg   : ONE 128-byte header block, then N text blocks.
 *
 * Header block fixed fields (ASCII, space-padded, PUBLISHED offsets):
 *   off   0 : status flag           1   (space = public unread)
 *   off   1 : message number        7   left-justified ASCII
 *   off   8 : date "MM-DD-YY"        8
 *   off  16 : time "HH:MM"           5
 *   off  21 : To                    25
 *   off  46 : From                  25
 *   off  71 : Subject               25
 *   off  96 : password             12   (spaces)
 *   off 108 : reference number      8   (spaces or "0")
 *   off 116 : block count           6   ASCII, INCLUDING the header block
 *   off 122 : active flag           1   (0xE1 = active)
 *   off 123 : conference number     2   little-endian u16
 *   off 125 : logical-message ind.  3   (spaces)
 *                                 ----
 *                                  128
 *
 * Body text: each '\n' stored as 0xE3; final block space-padded to 128.
 * blockcount = 1 + ceil(bodylen / 128).
 */
#include <stdio.h>
#include <string.h>
#include <time.h>
#include "qwk.h"

#define QWK_CR 0xE3   /* on-disk substitute for an end-of-line */

static FILE  *g_fp;
static pp_u32 g_count;

/* ---- field helpers ------------------------------------------------------ */

/* copy up to max bytes of s into p, then space-pad to max. */
static void put_field(pp_u8 *p, const char *s, int max)
{
    int i = 0;
    while (i < max && s[i]) { p[i] = (pp_u8)s[i]; i++; }
    while (i < max) { p[i] = (pp_u8)' '; i++; }
}

/* render an unsigned value as decimal into a max-wide LEFT-justified,
 * space-padded ASCII field. */
static void put_num(pp_u8 *p, pp_u32 v, int max)
{
    char tmp[12];
    int  n = 0;
    int  i;
    if (v == 0) {
        tmp[n++] = '0';
    } else {
        while (v > 0 && n < (int)sizeof tmp) { tmp[n++] = (char)('0' + (int)(v % 10)); v /= 10; }
    }
    /* tmp holds the digits reversed; emit them forwards, then pad. */
    for (i = 0; i < max; i++) {
        p[i] = (n > 0) ? (pp_u8)tmp[--n] : (pp_u8)' ';
    }
}

/* parse a left-justified ASCII unsigned decimal out of a fixed field. */
static pp_u32 get_num(const pp_u8 *p, int max)
{
    pp_u32 v = 0;
    int    i;
    for (i = 0; i < max; i++) {
        pp_u8 c = p[i];
        if (c >= '0' && c <= '9') v = v * 10 + (pp_u32)(c - '0');
        else if (c == ' ') continue;
        else break;
    }
    return v;
}

/* copy a fixed field, trim trailing spaces, NUL-terminate into out[outsz]. */
static void get_field(const pp_u8 *p, int max, char *out, int outsz)
{
    int n = (max < outsz - 1) ? max : outsz - 1;
    int i;
    for (i = 0; i < n; i++) out[i] = (char)p[i];
    while (i > 0 && (out[i - 1] == ' ' || out[i - 1] == '\0')) i--;
    out[i] = '\0';
}

static void put_u16le(pp_u8 *p, pp_u16 v)
{
    p[0] = (pp_u8)(v & 0xff);
    p[1] = (pp_u8)((v >> 8) & 0xff);
}

static pp_u16 get_u16le(const pp_u8 *p)
{
    return (pp_u16)(p[0] | (p[1] << 8));
}

/* ---- writing ------------------------------------------------------------ */

int qwk_begin(const char *msgpath, const char *board)
{
    pp_u8 blk[QWK_BLOCK];
    char  stamp[QWK_BLOCK + 1];
    int   n;

    g_count = 0;
    g_fp = fopen(msgpath, "wb");
    if (g_fp == (FILE *)0) return 1;

    /* "Produced by Vendetta/X" + the board name, the rest space-padded. */
    stamp[0] = '\0';
    strcpy(stamp, "Produced by Vendetta/X <");
    n = (int)strlen(stamp);
    if (board) {
        int i = 0;
        while (board[i] && n < QWK_BLOCK - 1) { stamp[n++] = board[i++]; }
    }
    if (n < QWK_BLOCK - 1) stamp[n++] = '>';
    stamp[n] = '\0';

    put_field(blk, stamp, QWK_BLOCK);
    if (fwrite(blk, 1, QWK_BLOCK, g_fp) != QWK_BLOCK) {
        fclose(g_fp); g_fp = (FILE *)0; return 1;
    }
    return 0;
}

int qwk_add(const pp_qwk_msg *m)
{
    pp_u8      hdr[QWK_BLOCK];
    pp_u8      txt[QWK_BLOCK];
    char       datebuf[16];
    char       timebuf[16];
    const char *body;
    long       bodylen;
    long       textblocks;
    pp_u32     blockcount;
    long       i;
    int        col;
    struct tm *tmv;
    time_t     t;

    if (g_fp == (FILE *)0 || m == (const pp_qwk_msg *)0) return 1;

    body = m->body ? m->body : "";
    bodylen = (long)strlen(body);
    textblocks = (bodylen + QWK_BLOCK - 1) / QWK_BLOCK;   /* ceil */
    if (bodylen == 0) textblocks = 0;
    blockcount = (pp_u32)(1 + textblocks);

    /* format date/time from the unix timestamp. */
    t = (time_t)m->when;
    tmv = gmtime(&t);
    if (tmv != (struct tm *)0) {
        sprintf(datebuf, "%02d-%02d-%02d",
                tmv->tm_mon + 1, tmv->tm_mday, (tmv->tm_year + 1900) % 100);
        sprintf(timebuf, "%02d:%02d", tmv->tm_hour, tmv->tm_min);
    } else {
        strcpy(datebuf, "01-01-80");
        strcpy(timebuf, "00:00");
    }

    /* ---- header block ---- */
    memset(hdr, ' ', QWK_BLOCK);
    hdr[0] = (pp_u8)' ';                            /* status: public unread */
    put_num(hdr + 1, m->number, 7);                 /* message number */
    put_field(hdr + 8,  datebuf, 8);                /* MM-DD-YY */
    put_field(hdr + 16, timebuf, 5);                /* HH:MM */
    put_field(hdr + 21, m->to,      QWK_TO_MAX);
    put_field(hdr + 46, m->from,    QWK_FROM_MAX);
    put_field(hdr + 71, m->subject, QWK_SUBJ_MAX);
    put_field(hdr + 96, "", 12);                    /* password (spaces) */
    put_field(hdr + 108, "0", 8);                   /* reference number */
    put_num(hdr + 116, blockcount, 6);              /* block count incl. header */
    hdr[122] = (pp_u8)0xE1;                          /* active */
    put_u16le(hdr + 123, m->conference);            /* conference number */
    hdr[125] = (pp_u8)' ';                          /* logical-message ind. */
    hdr[126] = (pp_u8)' ';
    hdr[127] = (pp_u8)' ';

    if (fwrite(hdr, 1, QWK_BLOCK, g_fp) != QWK_BLOCK) return 1;

    /* ---- text blocks ---- */
    col = 0;
    for (i = 0; i < bodylen; i++) {
        char c = body[i];
        txt[col++] = (c == '\n') ? (pp_u8)QWK_CR : (pp_u8)c;
        if (col == QWK_BLOCK) {
            if (fwrite(txt, 1, QWK_BLOCK, g_fp) != QWK_BLOCK) return 1;
            col = 0;
        }
    }
    if (col > 0) {                                  /* pad final block */
        while (col < QWK_BLOCK) txt[col++] = (pp_u8)' ';
        if (fwrite(txt, 1, QWK_BLOCK, g_fp) != QWK_BLOCK) return 1;
    }

    g_count++;
    return 0;
}

int qwk_finish(void)
{
    int n = (int)g_count;
    if (g_fp) {
        fflush(g_fp);
        fclose(g_fp);
        g_fp = (FILE *)0;
    }
    return n;
}

/* ---- CONTROL.DAT -------------------------------------------------------- */

int qwk_write_control(const char *path, const char *board, const char *location,
                      const char *sysop, const char **conf_names, int nconf)
{
    FILE      *f;
    int        i;
    int        maxconf;
    time_t     t;
    struct tm *tmv;
    char       datebuf[32];

    f = fopen(path, "wb");
    if (f == (FILE *)0) return 1;

    t = time((time_t *)0);
    tmv = gmtime(&t);
    if (tmv != (struct tm *)0) {
        sprintf(datebuf, "%02d-%02d-%04d,%02d:%02d:%02d",
                tmv->tm_mon + 1, tmv->tm_mday, tmv->tm_year + 1900,
                tmv->tm_hour, tmv->tm_min, tmv->tm_sec);
    } else {
        strcpy(datebuf, "01-01-1980,00:00:00");
    }

    /* Published CONTROL.DAT layout, CRLF lines:
     *   board name
     *   location (city, state)
     *   sysop phone (placeholder)
     *   "00000000,BBSID"  (serial + board id)
     *   sysop name
     *   date "MM-DD-YYYY,HH:MM:SS"
     *   caller (the QWK user) name
     *   "0"  (mail counter / placeholder)
     *   max conference number (count - 1)
     *   then per conference: number line, name line
     */
    fprintf(f, "%s\r\n", board ? board : "");
    fprintf(f, "%s\r\n", location ? location : "");
    fprintf(f, "000-000-0000\r\n");
    fprintf(f, "00000000,VENDETTAX\r\n");
    fprintf(f, "%s\r\n", sysop ? sysop : "");
    fprintf(f, "%s\r\n", datebuf);
    fprintf(f, "%s\r\n", "VENDETTAX");
    fprintf(f, "0\r\n");
    fprintf(f, "0\r\n");

    maxconf = (nconf > 0) ? nconf - 1 : 0;
    fprintf(f, "%d\r\n", maxconf);
    for (i = 0; i < nconf; i++) {
        fprintf(f, "%d\r\n", i);
        fprintf(f, "%s\r\n", (conf_names && conf_names[i]) ? conf_names[i] : "");
    }

    fflush(f);
    fclose(f);
    return 0;
}

/* ---- reading ------------------------------------------------------------ */

int qwk_read(const char *msgpath, qwk_msg_cb cb, void *user)
{
    static char body[8 * 1024];          /* reused body buffer */
    pp_u8       hdr[QWK_BLOCK];
    pp_u8       txt[QWK_BLOCK];
    FILE       *f;
    pp_qwk_msg  m;
    pp_u32      blockcount;
    long        textblocks;
    long        b;
    int         j;
    int         pos;
    int         count = 0;

    f = fopen(msgpath, "rb");
    if (f == (FILE *)0) return -1;

    /* skip block 0 */
    if (fread(hdr, 1, QWK_BLOCK, f) != QWK_BLOCK) { fclose(f); return -1; }

    for (;;) {
        size_t got = fread(hdr, 1, QWK_BLOCK, f);
        if (got == 0) break;                       /* clean EOF */
        if (got != QWK_BLOCK) { fclose(f); return -1; }

        blockcount = get_num(hdr + 116, 6);
        if (blockcount == 0) { fclose(f); return -1; }   /* malformed */
        textblocks = (long)blockcount - 1;

        memset(&m, 0, sizeof m);
        m.number     = get_num(hdr + 1, 7);
        m.conference = get_u16le(hdr + 123);
        get_field(hdr + 21, QWK_TO_MAX,   m.to,      (int)sizeof m.to);
        get_field(hdr + 46, QWK_FROM_MAX, m.from,    (int)sizeof m.from);
        get_field(hdr + 71, QWK_SUBJ_MAX, m.subject, (int)sizeof m.subject);

        /* read text blocks into the reused buffer */
        pos = 0;
        for (b = 0; b < textblocks; b++) {
            if (fread(txt, 1, QWK_BLOCK, f) != QWK_BLOCK) { fclose(f); return -1; }
            for (j = 0; j < QWK_BLOCK; j++) {
                char c = (txt[j] == (pp_u8)QWK_CR) ? '\n' : (char)txt[j];
                if (pos < (int)sizeof body - 1) body[pos++] = c;
            }
        }
        /* trim trailing ^Z / space padding (but keep interior newlines). */
        while (pos > 0 && (body[pos - 1] == ' ' || body[pos - 1] == 0x1A ||
                           body[pos - 1] == '\n' || body[pos - 1] == '\0')) {
            pos--;
        }
        body[pos] = '\0';
        m.body = body;

        if (cb) cb(user, &m);
        count++;
    }

    fclose(f);
    return count;
}
